File: app/src/main/java/com/example/sunkai/heritage/personFragment.java
package com.example.sunkai.heritage;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.BroadcastReceiver;
import android.content.ContentUris;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.provider.DocumentsContract;
import android.provider.MediaStore;
import android.support.v4.app.Fragment;
import android.support.v4.content.FileProvider;
import android.support.v7.app.AlertDialog;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.sunkai.heritage.ConnectWebService.HandlePerson;
import com.example.sunkai.heritage.Data.HandlePic;
import com.makeramen.roundedimageview.RoundedImageView;

import org.kobjects.base64.Base64;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.InputStream;
import java.lang.reflect.Field;

/**
 * 此类是用于处理个人中心页面
 */


public class personFragment extends Fragment implements View.OnClickListener{
    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";

    // TODO: Rename and change types of parameters
    private String mParam1;
    private String mParam2;
    private LinearLayout orderLinear,myOwnTiezi;
    private TextView userName,follow,followNumber,fans,fansNumber;
    private RoundedImageView userImage;

    AlertDialog.Builder builder;
    AlertDialog ad;

    Uri outputUri;//剪裁后的图片的uri


    private static final int IS_INTO_LOGIN=1;
    private static final int REQUEST_PICK_IMAGE = 11; //相册选取
    private static final int REQUEST_CAPTURE = 12;  //拍照
    private static final int REQUEST_PICTURE_CUT = 13;  //剪裁图片
    private static final int REQUEST_PERMISSION = 14;  //权限请求
    private PermissionsChecker mPermissionsChecker; // 权限检测器
    static final String[] PERMISSIONS = new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.CAMERA};
    private Uri imageUri;//原图保存地址
    private boolean isClickCamera;
    private String imagePath;
    public personFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @param param2 Parameter 2.
     * @return A new instance of fragment personFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static personFragment newInstance(String param1, String param2) {
        personFragment fragment = new personFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
            mParam2 = getArguments().getString(ARG_PARAM2);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        setHasOptionsMenu(true);
        View view=inflater.inflate(R.layout.fragment_person, container, false);
        orderLinear=(LinearLayout)view.findViewById(R.id.fragment_person_oder_linner);
        myOwnTiezi=(LinearLayout)view.findViewById(R.id.fragment_person_my_tiezi);
        orderLinear.setOnClickListener(this);
        myOwnTiezi.setOnClickListener(this);
        userName=(TextView)view.findViewById(R.id.sign_name_textview);
        if(LoginActivity.userName==null){
            userName.setText("没有登录");
        }
        else{
            userName.setText(LoginActivity.userName);
        }
        follow=(TextView)view.findViewById(R.id.person_follow);
        followNumber=(TextView)view.findViewById(R.id.person_follow_number);
        fans=(TextView)view.findViewById(R.id.person_fans);
        fansNumber=(TextView)view.findViewById(R.id.person_fans_number);
        userImage=(RoundedImageView)view.findViewById(R.id.sign_in_icon);
        follow.setOnClickListener(this);
        followNumber.setOnClickListener(this);
        fans.setOnClickListener(this);
        fansNumber.setOnClickListener(this);
        userImage.setOnClickListener(this);
        mPermissionsChecker = new PermissionsChecker(getActivity());
        GetUserInfo();
        IntentFilter intentFilter=new IntentFilter();
        intentFilter.addAction("android.intent.action.focusAndFansCountChange");
        getActivity().registerReceiver(focusAndFansCountChange,intentFilter);
        intentFilter=new IntentFilter();
        intentFilter.addAction("android.intent.action.refreInfomation");
        getActivity().registerReceiver(refreshInfo,intentFilter);
        return view;
    }

    /**
     * 开始获取用户的粉丝和关注的信息
     * 先获取关注信息
     */
    private void GetUserInfo(){
        new Thread(getFollowCount).start();
    }
    private void toLogin(){
        Toast.makeText(getActivity(),"没有登录",Toast.LENGTH_SHORT).show();
        Intent intent=new Intent(getActivity(),LoginActivity.class);
        intent.putExtra("isInto",1);
        startActivityForResult(intent,1);
    }

    @Override
    public void onClick(View v){
        Intent intent;
        if(LoginActivity.userID==0){
            toLogin();
            return;
        }
        switch (v.getId()){
            case R.id.fragment_person_oder_linner:
                intent=new Intent(getActivity(),MyOrderActivity.class);
                startActivity(intent);
                break;
            case R.id.fragment_person_my_tiezi:
                intent=new Intent(getActivity(),UserOwnTiezi.class);
                startActivity(intent);
                break;
            case R.id.person_follow:
            case R.id.person_follow_number:
                intent=new Intent(getActivity(),focusInformation.class);
                intent.putExtra("information","focus");
                startActivity(intent);
                break;
            case R.id.person_fans:
            case R.id.person_fans_number:
                intent=new Intent(getActivity(),focusInformation.class);
                intent.putExtra("information","fans");
                startActivity(intent);
                break;
            case R.id.sign_in_icon:
                builder=new AlertDialog.Builder(getActivity()).setTitle("请选择");
                final ListView choice=new ListView(getActivity());
                ArrayAdapter arrayAdapter=new ArrayAdapter<String>(getActivity(),android.R.layout.simple_list_item_1);
                arrayAdapter.add("拍照");
                arrayAdapter.add("从图库中选择");
                final android.app.AlertDialog dialog;
                choice.setAdapter(arrayAdapter);
                builder.setView(choice);
                ad = builder.show();

                /**
                 * 在获取图片或者拍照的时候，判断系统版本，如果是安卓6.0以上，要动态申请权限
                 */
                choice.setOnItemClickListener((parent, view, position, id) -> {
                    if(position==0)
                    {
                        ad.dismiss();
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                            if (mPermissionsChecker.lacksPermissions(PERMISSIONS)) {
                                startPermissionsActivity();
                            } else {
                                openCamera();
                            }
                        } else {
                            openCamera();
                        }
                        isClickCamera = true;
                    }
                    else if(position==1)
                    {
                        ad.dismiss();
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                            if (mPermissionsChecker.lacksPermissions(PERMISSIONS)) {
                                startPermissionsActivity();
                            } else {
                                selectFromAlbum();
                            }
                        } else {
                            selectFromAlbum();
                        }
                        isClickCamera = false;
                    }
                });
                break;
            default:
                break;
        }
    }
    /**
     * 打开系统相机
     */
    private void openCamera() {
        File file = new FileStorage().createIconFile();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            imageUri = FileProvider.getUriForFile(getActivity(), "com.example.sunkai.heritage.fileprovider", file);//通过FileProvider创建一个content类型的Uri
        } else {
            imageUri = Uri.fromFile(file);
        }
        Intent intent = new Intent();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION); //添加这一句表示对目标应用临时授权该Uri所代表的文件
        }
        intent.setAction(MediaStore.ACTION_IMAGE_CAPTURE);//设置Action为拍照
        intent.putExtra(MediaStore.EXTRA_OUTPUT, imageUri);//将拍取的照片保存到指定URI
        startActivityForResult(intent, REQUEST_CAPTURE);
    }

    /**
     * 从相册选择
     */
    private void selectFromAlbum() {
        Intent intent = new Intent(Intent.ACTION_PICK);
        intent.setDataAndType(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, "image/*");
        startActivityForResult(intent, REQUEST_PICK_IMAGE);
    }


//    /**
//     * 裁剪
//     */
//    private void cropPhoto() {
//        File file = new FileStorage().createCropFile();
//        Uri outputUri = Uri.fromFile(file);//缩略图保存地址
//        Intent intent = new Intent("com.android.camera.action.CROP");
//        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
//            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
//        }
//        intent.setDataAndType(imageUri, "image/*");
//        intent.putExtra("crop", "true");
//        intent.putExtra("aspectX", 1);
//        intent.putExtra("aspectY", 1);
//        intent.putExtra("scale", true);
//        intent.putExtra("return-data", false);
//        intent.putExtra(MediaStore.EXTRA_OUTPUT, outputUri);
//        intent.putExtra("outputFormat", Bitmap.CompressFormat.JPEG.toString());
//        intent.putExtra("noFaceDetection", true);
//        startActivityForResult(intent, REQUEST_PICTURE_CUT);
//    }

    private void startPermissionsActivity() {
        PermissionsActivity.startActivityForResult(getActivity(), REQUEST_PERMISSION,
                PERMISSIONS);
    }

    @TargetApi(19)
    private void handleImageOnKitKat(Intent data) {
        imagePath = null;
        if(data==null){
            return;
        }
        imageUri = data.getData();
        if (DocumentsContract.isDocumentUri(getActivity(), imageUri)) {
            //如果是document类型的uri,则通过document id处理
            String docId = DocumentsContract.getDocumentId(imageUri);
            if ("com.android.providers.media.documents".equals(imageUri.getAuthority())) {
                String id = docId.split(":")[1];//解析出数字格式的id
                String selection = MediaStore.Images.Media._ID + "=" + id;
                imagePath = getImagePath(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, selection);
            } else if ("com.android.downloads.documents".equals(imageUri.getAuthority())) {
                Uri contentUri = ContentUris.withAppendedId(Uri.parse("content://downloads/public_downloads"), Long.valueOf(docId));
                imagePath = getImagePath(contentUri, null);
            }
        } else if ("content".equalsIgnoreCase(imageUri.getScheme())) {
            //如果是content类型的Uri，则使用普通方式处理
            imagePath = getImagePath(imageUri, null);
        } else if ("file".equalsIgnoreCase(imageUri.getScheme())) {
            //如果是file类型的Uri,直接获取图片路径即可
            imagePath = imageUri.getPath();
        }

        cropPhoto();
//        setImage();
    }

    private void handleImageBeforeKitKat(Intent intent) {
        imageUri = intent.getData();
        imagePath = getImagePath(imageUri, null);
        cropPhoto();
//        setImage();
    }

    private String getImagePath(Uri uri, String selection) {
        String path = null;
        //通过Uri和selection老获取真实的图片路径
        Cursor cursor = getActivity().getContentResolver().query(uri, null, selection, null, null);
        if (cursor != null) {
            if (cursor.moveToFirst()) {
                path = cursor.getString(cursor.getColumnIndex(MediaStore.Images.Media.DATA));
            }
            cursor.close();
        }
        return path;
    }

    private void setImage(){
        Bitmap bitmap = null;
        try {
            /**
             * 在经过剪裁之后，outputUri会被赋值
             * 将uri指向的内容复制给bitmap
             */
            bitmap = BitmapFactory.decodeStream(getActivity().getContentResolver().openInputStream(outputUri));
            userImage.setImageDrawable(null);
            System.gc();
            bitmap=HandlePic.compressBitmapToFile(bitmap,192,192);
            userImage.setImageBitmap(bitmap);
//            System.out.println("孙楷最帅"+bitmap.getWidth()+","+bitmap.getHeight());
            builder=new AlertDialog.Builder(getActivity()).setTitle("上传中").setView(R.layout.update_image_builder);
            ad=builder.create();
            ad.show();
            new Thread(updateUserImage).start();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
        /**
     * 裁剪
     */

    private void cropPhoto() {
        File file = new FileStorage().createCropFile();
        outputUri = Uri.fromFile(file);//缩略图保存地址
        Intent intent = new Intent("com.android.camera.action.CROP");
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        }
        intent.setDataAndType(imageUri, "image/*");
        intent.putExtra("crop", "true");
        intent.putExtra("aspectX", 1);
        intent.putExtra("aspectY", 1);
        intent.putExtra("scale", true);
        intent.putExtra("return-data", false);
        intent.putExtra(MediaStore.EXTRA_OUTPUT, outputUri);//剪裁输出的uri
        intent.putExtra("outputFormat", Bitmap.CompressFormat.JPEG.toString());
        intent.putExtra("noFaceDetection", false);//人脸检测
        startActivityForResult(intent, REQUEST_PICTURE_CUT);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case IS_INTO_LOGIN:
                userName.setText(LoginActivity.userName);
                GetUserInfo();
                break;
            case REQUEST_PICK_IMAGE://从相册选择
                if (Build.VERSION.SDK_INT >= 19) {
                    handleImageOnKitKat(data);
                } else {
                    handleImageBeforeKitKat(data);
                }
                break;
            case REQUEST_CAPTURE://拍照
                if (resultCode == getActivity().RESULT_OK) {
//                    cropPhoto();
                    cropPhoto();
                }
                break;
            case REQUEST_PICTURE_CUT://裁剪完成
                setImage();
                break;
            case REQUEST_PERMISSION://权限请求
                if (resultCode == PermissionsActivity.PERMISSIONS_DENIED) {
                    getActivity().finish();
                } else {
                    if (isClickCamera) {
                        openCamera();
                    } else {
                        selectFromAlbum();
                    }
                }
                break;
        }
    }
    @Override
    public void onDetach() {
        super.onDetach();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        getActivity().unregisterReceiver(focusAndFansCountChange);
        getActivity().unregisterReceiver(refreshInfo);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        Intent intent;
        switch (item.getItemId()){
            case R.id.action_search_user:
                intent=new Intent(getActivity(),SearchActivity.class);
                startActivity(intent);
                break;
            case R.id.action_setting:
                if(LoginActivity.userID==0){
                    toLogin();
                    break;
                }
                intent=new Intent(getActivity(),SettingActivity.class);
                userImage.setDrawingCacheEnabled(true);
                intent.putExtra("userImage",userImage.getDrawingCache());
                startActivity(intent);
        }
        return super.onOptionsItemSelected(item);
    }

    BroadcastReceiver focusAndFansCountChange=new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if("change".equals(intent.getStringExtra("message"))){
                new Thread(getFollowCount).start();
            }
            if("sigh_out".equals(intent.getStringExtra("message"))){
                intent=new Intent(getActivity(),LoginActivity.class);
                startActivity(intent);
                getActivity().finish();
            }
        }
    };

    Runnable getFollowCount=new Runnable() {
        @Override
        public void run() {
            int count= HandlePerson.Get_Follow_Number(LoginActivity.userID);
            getFollowCountHandler.sendEmptyMessage(count);
        }
    };
    Handler getFollowCountHandler=new Handler(){
        @Override
        public void handleMessage(Message msg){
            followNumber.setText(String.valueOf(msg.what));
            new Thread(getFansCount).start();
        }
    };
    Runnable getFansCount=new Runnable() {
        @Override
        public void run() {
            int count=HandlePerson.Get_Fans_Number(LoginActivity.userID);
            getFanseCountHandler.sendEmptyMessage(count);
        }
    };
    Handler getFanseCountHandler=new Handler(){
        @Override
        public void handleMessage(Message msg){
            fansNumber.setText(String.valueOf(msg.what));
            new Thread(getUserImage).start();
        }
    };
    Runnable getUserImage=new Runnable() {
        @Override
        public void run() {
            String result=HandlePerson.Get_User_Image(LoginActivity.userID);
            if("Error".equals(result)){
                getUserImageHandler.sendEmptyMessage(0);
            }
            else{
                Message msg=new Message();
                Bundle data=new Bundle();
                data.putString("image",result);
                msg.setData(data);
                msg.what=1;
                getUserImageHandler.sendMessage(msg);
            }
        }
    };
    Handler getUserImageHandler=new Handler(){
        @Override
        public void handleMessage(Message msg) {
            if(msg.what==1){
                Bundle data=msg.getData();
                String imageCode=data.getString("image");
                byte[] imageByte= Base64.decode(imageCode);
                InputStream in=new ByteArrayInputStream(imageByte);
                Bitmap bitmap= HandlePic.handlePic(getActivity(),in,0);
                userImage.setImageBitmap(bitmap);
            }
        }
    };


    Runnable updateUserImage=new Runnable() {
        @Override
        public void run() {
            userImage.setDrawingCacheEnabled(true);
            Bitmap bitmap=userImage.getDrawingCache();
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            bitmap.compress(Bitmap.CompressFormat.JPEG,40,baos);
            byte[] image=baos.toByteArray();
            String imageCode=Base64.encode(image);
            boolean result=HandlePerson.Update_User_Image(LoginActivity.userID,imageCode);
            if(result){
                updateUserImageHandler.sendEmptyMessage(1);
            }
            else{
                updateUserImageHandler.sendEmptyMessage(0);
            }
        }
    };

    Handler updateUserImageHandler=new Handler(){
        @Override
        public void handleMessage(Message msg) {
            ad.dismiss();
            if(msg.what==1){
                Toast.makeText(getActivity(),"更新头像成功",Toast.LENGTH_SHORT).show();
            }
            else{
                Toast.makeText(getActivity(),"出现错误，请稍后再试",Toast.LENGTH_SHORT).show();
                userImage.setImageResource(R.drawable.ic_assignment_ind_deep_orange_200_48dp);
            }
        }
    };
    BroadcastReceiver refreshInfo=new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            userName.setText(LoginActivity.userName);
            GetUserInfo();
        }
    };
}

File: app/src/main/java/com/example/sunkai/heritage/Fragment/PersonFragment.kt
package com.example.sunkai.heritage.Fragment

import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Bitmap
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.menu.MenuBuilder
import android.view.*
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.ActionMenuView
import androidx.core.content.ContextCompat
import androidx.core.view.children
import androidx.core.view.forEach
import androidx.core.view.get
import com.bumptech.glide.request.RequestOptions
import com.example.sunkai.heritage.Activity.*
import com.example.sunkai.heritage.Activity.LoginActivity.LoginActivity
import com.example.sunkai.heritage.ConnectWebService.HandlePerson
import com.example.sunkai.heritage.Data.HandlePic
import com.example.sunkai.heritage.Dialog.ChangePasswordDialog
import com.example.sunkai.heritage.Dialog.NormalWarningDialog
import com.example.sunkai.heritage.Fragment.BaseFragment.BaseTakePhotoLazyLoadFragment
import com.example.sunkai.heritage.R
import com.example.sunkai.heritage.tools.MakeToast.toast
import com.example.sunkai.heritage.tools.getDarkThemeColor
import com.example.sunkai.heritage.tools.getLightThemeColor
import com.example.sunkai.heritage.tools.getThemeColor
import com.example.sunkai.heritage.value.*
import kotlinx.android.synthetic.main.fragment_person.*
import kotlinx.android.synthetic.main.user_view.*

/**
 * 此类是用于处理个人中心页面
 */


class PersonFragment : BaseTakePhotoLazyLoadFragment(), View.OnClickListener {

    private lateinit var builder: AlertDialog.Builder
    private lateinit var ad: AlertDialog

    //记录用户头像的url
    private var userImageUrl: String? = null


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_person, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (savedInstanceState == null) {
            initview()
        }
        if (LoginActivity.userName == null) {
            sign_name_textview.text = "没有登录"
        } else {
            sign_name_textview.text = LoginActivity.userName
        }
    }

    override fun onRestoreFragmentLoadInformation() {
        initview()
        lazyLoad()
    }


    override fun setNeedChangeThemeColorWidget() {
        changeThemeWidge.add(R.id.fragment_person_my_tiezi)
        changeThemeWidge.add(R.id.fragment_person_my_like)
        changeThemeWidge.add(R.id.fragment_person_my_collect)
        changeThemeWidge.add(R.id.fragment_person_my_message)
        changeThemeWidge.add(R.id.person_fragment_setting)
    }

    override fun changeWidgeTheme() {
        super.changeWidgeTheme()
        val colorArray = IntArray(2)
        colorArray[0] = getThemeColor()
        colorArray[1] = getLightThemeColor()
        val drawable = GradientDrawable(GradientDrawable.Orientation.TOP_BOTTOM, colorArray)
        userViewBackGround.setImageDrawable(drawable)
        sign_name_textview.setTextColor(getDarkThemeColor())
        person_follow.setTextColor(getDarkThemeColor())
        person_fans.setTextColor(getDarkThemeColor())
        person_fans_number.setTextColor(getDarkThemeColor())
        person_follow_number.setTextColor(getDarkThemeColor())
    }

    private fun initview() {
        setClick(fragmentPersonLinearLayout)
        val activity = activity
        activity?.let {
            if (activity is AppCompatActivity) {
                fragment_person_toolbar.title = ""
                fragment_person_toolbar.popupTheme
                activity.setSupportActionBar(fragment_person_toolbar)
                setHasOptionsMenu(true)
            }
        }
    }

    private fun setClick(viewGroup: ViewGroup) {
        viewGroup.forEach { if (it is ViewGroup) setClick(it) else it.setOnClickListener(this) }
    }

    override fun startLoadInformation() {
        checkLogin()
        GetUserInfo()
    }

    //获取用户的信息
    private fun GetUserInfo() {
        requestHttp {
            val userInfo = HandlePerson.GetUserAllInfo(LoginActivity.userID) ?: return@requestHttp
            val userImageUrl = HandlePerson.GetUserImageURL(LoginActivity.userID)
            val activity = activity ?: return@requestHttp
            activity.runOnUiThread {
                sign_name_textview.text = userInfo.userName
                person_fans_number.text = userInfo.fansNumber.toString()
                person_follow_number.text = userInfo.focusNumber.toString()
                if (userImageUrl != null) {
                    this.userImageUrl = userImageUrl
                    val requestOption = RequestOptions().placeholder(R.drawable.ic_assignment_ind_deep_orange_200_48dp).error(R.drawable.ic_assignment_ind_deep_orange_200_48dp)
                    glide.load(userImageUrl).apply(requestOption).into(sign_in_icon)
                }
            }
        }
    }

    private fun UpdateUserImage(bitmap: Bitmap) {
        Thread {
            val result = HandlePerson.UpdateUserImage(LoginActivity.userID, HandlePic.bitmapToByteArray(bitmap))
            activity?.runOnUiThread {
                ad.dismiss()
                if (result) {
                    Toast.makeText(activity, "更新头像成功", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(activity, "出现错误，请稍后再试", Toast.LENGTH_SHORT).show()
                    sign_in_icon.setImageResource(R.drawable.ic_assignment_ind_deep_orange_200_48dp)
                }
            }
        }.start()

    }


    override fun onClick(v: View) {
        if (checkLogin()) {
            val intent: Intent
            when (v.id) {
                R.id.fragment_person_my_tiezi -> {
                    intent = Intent(activity, UserOwnCommentActivity::class.java)
                    startActivity(intent)
                }
                R.id.person_fragment_setting -> {
                    if (checkLogin()) {
                        intent = Intent(activity, SettingListActivity::class.java)
                        startActivity(intent)
                    }
                }
                R.id.person_follow, R.id.person_follow_number -> {
                    intent = Intent(activity, FocusInformationActivity::class.java)
                    intent.putExtra(INFORMATION, FROM_FOCUS)
                    startActivityForResult(intent, FROM_FOCUS_AND_FANS_INFORMATION)
                }
                R.id.person_fans, R.id.person_fans_number -> {
                    intent = Intent(activity, FocusInformationActivity::class.java)
                    intent.putExtra(INFORMATION, FROM_FANS)
                    startActivityForResult(intent, FROM_FOCUS_AND_FANS_INFORMATION)
                }
                R.id.fragment_person_my_like -> {
                    intent = Intent(activity, UserLikeCommentActivity::class.java)
                    startActivity(intent)
                }
                R.id.fragment_person_my_collect -> {
                    intent = Intent(activity, MyCollectionActivity::class.java)
                    startActivity(intent)
                }
                R.id.fragment_person_my_message -> {
                    intent = Intent(activity, MyMessageActivity::class.java)
                    startActivity(intent)
                }
            }
        }
    }

    override fun getNeedOpenChooseImageView(): Array<View> {
        return arrayOf(sign_in_icon)
    }

    //设置用户头像的回调
    override fun setImageToImageView(bitmap: Bitmap) {
        val compressBitmap = HandlePic.compressBitmapToFile(bitmap, 192, 192)
        sign_in_icon.setImageBitmap(compressBitmap)
        builder = AlertDialog.Builder(activity!!).setTitle("上传中").setView(R.layout.update_image_builder)
        ad = builder.create()
        ad.show()
        UpdateUserImage(compressBitmap)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        menu?.clear()
        inflater?.inflate(R.menu.person_fragment_menu, menu)

    }

    //通过反射来让弹出的菜单显示图标
    override fun onPrepareOptionsMenu(menu: Menu?) {
        menu?.let {
            if (menu.javaClass == MenuBuilder::class.java) {
                try {
                    val m = menu.javaClass.getDeclaredMethod("setOptionalIconsVisible", Boolean::class.java)
                    m.isAccessible = true
                    m.invoke(menu, true)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
        super.onPrepareOptionsMenu(menu)
    }


    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.log_out -> {
                sign_out()
            }
            R.id.change_password -> {
                changePassword()
            }
            R.id.search_user -> {
                searchUser()
            }
            R.id.about_us -> {
                showAboutUs()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            IS_INTO_LOGIN -> {
                sign_name_textview.text = LoginActivity.userName
                GetUserInfo()
            }
            SETTING_ACTIVITY -> {
                if (resultCode == SIGN_OUT) {
                    checkLogin()
                    activity?.finish()
                }
            }
            FROM_FOCUS_AND_FANS_INFORMATION, FROM_SEARCH_ACTIVITY -> {
                if (resultCode == STATE_CHANGE) {
                    GetUserInfo()
                }
            }
        }
    }

    private fun sign_out() {
        NormalWarningDialog().setTitle("是否注销?")
                .setOnSubmitClickListener(object : NormalWarningDialog.onSubmitClickListener {
                    override fun onSubmit(view: View, dialog: NormalWarningDialog) {
                        activity?.getSharedPreferences("data", Context.MODE_PRIVATE)?.edit()?.clear()?.apply()//清除自动登录的信息
                        LoginActivity.userID = 0
                        LoginActivity.userName = null
                        activity?.finish()
                        checkLogin()
                    }
                })
                .setSubmitText("退出")
                .show(activity?.supportFragmentManager, "exitDialog")
    }

    private fun changePassword() {
        if (checkLogin()) {
            val activity = activity
            activity?.let {
                val dialog = ChangePasswordDialog()
                dialog.show(fragmentManager, "修改密码")
            }
        }
    }

    private fun searchUser() {
        val intent = Intent(activity, SearchActivity::class.java)
        intent.putExtra(SEARCH_TYPE, TYPE_USER)
        startActivityForResult(intent, FROM_SEARCH_ACTIVITY)
    }

    private fun showAboutUs() {
        val intent = Intent(activity, AboutUSActivity::class.java)
        startActivity(intent)
    }

    private fun checkLogin(): Boolean {
        if (LoginActivity.userID == 0) {
            toast("没有登录")
            val intent = Intent(activity, LoginActivity::class.java)
            intent.putExtra("isInto", LOG_OUT)
            startActivityForResult(intent, 1)
            return false
        }
        return true
    }


    companion object {
        private const val IS_INTO_LOGIN = 1
        private const val FROM_FOCUS_AND_FANS_INFORMATION = 2
        private const val FROM_SEARCH_ACTIVITY = 3
    }
}
